### PR TITLE
Feature/40 display whispers as designed

### PR DESCRIPTION
**Pull request by wayangalihpratama**
### TODO / DONE

---

- [x] Update ai message design, add copy functionality & closeable
- [x] Refine socketio event to parent chat page
- [x] Update chat list for incoming message
- [x] Implement react md render
- [x] Allow to use whisper as reply message prefill
- [x] Handle `\n` in markdown renderer

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208050895370739
  - https://app.asana.com/0/0/1207949534161317
  - https://app.asana.com/0/0/1207949534161316

### CODE REVIEW

File: frontend/src/app/chats/page.js [hunk 1]
```diff
@@ -1,34 +1,176 @@
...
+      socket.off("chats", onChats);
+      socket.off("whisper", onWhisper);
       socket.off("disconnect", onDisconnect);
     };
   });
 
-  return clientPhoneNumber ? <ChatWindow /> : <ChatList />;
+  // handle on click notification
+  const handleOnClickNotification = (sender) => {
+    const findClient = clients.find((c) => c.phone_number === sender);
```
Comment (dedenbangkit) [inline frontend/src/app/chats/page.js:104]: const name should be `selectedClient`

File: frontend/src/components/chats/ChatWindow.js [hunk 2]
```diff
@@ -133,88 +151,51 @@ const ChatWindow = () => {
...
-                  </Fragment>
-                ))}
-              </p>
-              <p className="text-right text-xs text-gray-400 mt-2">10:00 AM</p>
-            </div>
-          </div>
-        );
-      }
-    });
+  const renderChats = () => {
```
Comment (dedenbangkit) [inline frontend/src/components/chats/ChatWindow.js:173]: can we do `useMemo` for this render?

File: frontend/src/components/chats/ChatWindow.js [hunk 3]
```diff
@@ -261,40 +242,28 @@ const ChatWindow = () => {
           className="flex-1 p-4 overflow-auto border-b"
         >
           {renderChatHistory()}
-          {renderChatMessages()}
+          {renderChats()}
```
Comment (dedenbangkit) [inline frontend/src/components/chats/ChatWindow.js:245]: useMemo

File: frontend/src/app/chats/page.js [hunk 4]
```diff
@@ -1,34 +1,176 @@
...
+        setReloadChatList(!findClient);
+
+        // to handle show & loading whisper
+        setShowWhisper((prev) => [
+          ...new Set([
+            ...prev,
+            value.conversation_envelope.client_phone_number,
+          ]),
+        ]);
+        setLoadingWhisper((prev) => [
```
Comment (dedenbangkit) [inline frontend/src/app/chats/page.js:47]: Can we simplify to something like this? `[{"conversationId": 1, "loading": true, "message": "something"}]` so setLoading can be: `{...prev.filter((x) => x.conversationId !== conversationId), {conversationId: new,  "message": null}`. What do you think?

**Review by dedenbangkit — CHANGES_REQUESTED** [review]

File: frontend/src/app/chats/page.js (hunk 4 above)
Comment (wayangalihpratama) [reply frontend/src/app/chats/page.js:47]: Yes, I think we can simplify that, instead of using multiple state.

Comment (dedenbangkit) [reply frontend/src/app/chats/page.js:47]: mantap

**Review by dedenbangkit — APPROVED** [review]

**Pull request merged**